### PR TITLE
[ProjectConverter] Fix outputs keying to prevent cache missing, duplicates

### PR DESCRIPTION
**Pull request by FredKSchott**
*A part of #217: "Refactor ProjectConverter to Fix URL Handling"*
*First PR! Start here!*

## Summary:

Previously, `outputs` contained a mishmash of url. Some used the old bower file location, some used the new npm file location, and some were manually constructed. This has caused cache misses and slower conversions (and annoying bugs as I've been poking around and refactoring).

This PR fixes all of these issues by always keying to the OriginalDocumentUrl, along with some naming clarifications.

### CODE REVIEW

File: src/tools/test-polymer.ts [hunk 1]
```diff
@@ -75,7 +75,7 @@ function rework(line: string) {
                           .filter(([_, v]) => v !== undefined)
                           .map(([k]) => k);
   const expectedPaths = IterableX.from(walkDir(expectedDir))
-                            .map((f) => `./${f}` as ConvertedDocumentUrl)
+                            .map((f) => `${f}` as ConvertedDocumentFilePath)
```
Comment (usergenic) [inline src/tools/test-polymer.ts]: Immaterial, but style-wise it feels like, for consistency, this should be:
```ts
.map((f) => f as string as ConvertedDocumentFilePath)
```

Comment (FredKSchott) [reply src/tools/test-polymer.ts]: good catch, no longer need the template string since there's no more string modifications

**Comment by FredKSchott** [comment]
Thanks @usergenic! @justinfagnani PTAL

File: src/document-converter.ts [hunk 2]
```diff
@@ -235,24 +235,22 @@ export class DocumentConverter {
...
         recast.print(program, {quote: 'single', wrapColumn: 80, tabWidth: 2});
-    const deletedHtmlFileUrl =
-        ('./' + this.originalUrl) as ConvertedDocumentUrl;
-    return [
-      {
+
+    return {
+      originalUrl: this.originalUrl,
+      convertedUrl: this.convertedUrl,
+      convertedFilePath: this.originalUrl.replace('.html', '.js') as
```
Comment (justinfagnani) [inline src/document-converter.ts]: This looks potentially dangerous due to the unrestricted replace. Can you move to a central method to convert .html URLs to .js, and maybe only replace extensions?

File: src/document-converter.ts [hunk 3]
```diff
@@ -182,10 +182,10 @@ export class DocumentConverter {
         });
   }
 
-  convertToJsModule(): Iterable<ConversionOutput> {
+  convertToJsModule(): ConversionResult|void {
     for (const exclude of this.conversionMetadata.excludes) {
       if (this.originalUrl.endsWith(exclude)) {
-        return [];
+        return;
```
Comment (justinfagnani) [inline src/document-converter.ts]: Would it make sense to return a no-op ConversionResult? You use `results.has()` as a gate above, and populating the map would prevent multiple visits to this method.

File: src/document-converter.ts [hunk 4]
```diff
@@ -404,11 +402,17 @@ export class DocumentConverter {
...
-    return [{
-      type: 'html-file',
-      url: ('./' + this.originalUrl) as ConvertedDocumentUrl,
-      source: contents,
-    }];
+
+    return {
+      originalUrl: this.originalUrl,
+      convertedUrl: this.convertedUrl,
+      convertedFilePath: this.originalUrl as string as
```
Comment (justinfagnani) [inline src/document-converter.ts]: Consider making a no-op conversion method that just does the cast.

File: src/js-module.ts [hunk 5]
```diff
@@ -12,42 +12,31 @@
...
+  /** Set of exported names. */
   readonly es6Exports: ReadonlySet<string>;
+}
 
-  readonly exportedNamespaceMembers: ReadonlyArray<NamespaceMemberToExport>;
+export interface ConversionResult {
+  readonly originalUrl: OriginalDocumentUrl;
+  readonly convertedUrl: ConvertedDocumentUrl;
+  readonly convertedFilePath: ConvertedDocumentFilePath;
+  readonly keepOriginal?: boolean;
```
Comment (justinfagnani) [inline src/js-module.ts]: Can you document this one? What are the implied side effects of this being true? What's the intended default?

File: src/base-converter.ts [hunk 6]
```diff
@@ -112,63 +113,65 @@ export abstract class BaseConverter {
...
-  private handleNewJsModules(outputs: Iterable<ConversionOutput>): void {
-    for (const output of outputs) {
-      this.outputs.set(output.url, output);
-      if (output.type === 'js-module') {
-        for (const expr of output.exportedNamespaceMembers) {
-          this.namespacedExports.set(
-              expr.oldNamespacedName,
-              new JsExport(output.url, expr.es6ExportName));
-        }
+  private handleNewJsModules(newModule: ConversionResult): void {
```
[the remark this answers is not in the document]
Comment (FredKSchott) [reply src/base-converter.ts]: This interface is superseded by `UrlHandler`: #224

File: src/document-converter.ts (hunk 2 above)
Comment (FredKSchott) [reply src/document-converter.ts]: Good idea, I must have missed this when I was looking for other instances of fixHtmlExtension. Moving this into a helper method with the correct implementation.

File: src/base-converter.ts [hunk 7]
```diff
@@ -112,63 +113,65 @@ export abstract class BaseConverter {
...
    */
   convertHtmlToHtml(document: Document, visited: Set<OriginalDocumentUrl>) {
-    const htmlUrl = `./${document.url}` as ConvertedDocumentUrl;
-    if (!this.outputs.has(htmlUrl)) {
-      const newModules = this.getDocumentConverter(document, visited)
-                             .convertAsToplevelHtmlDocument();
-      this.handleNewJsModules(newModules);
+    if (!this.results.has(getDocumentUrl(document))) {
+      const newModule = this.getDocumentConverter(document, visited)
+                            .convertAsToplevelHtmlDocument();
```
[the remark this answers is not in the document]
Comment (FredKSchott) [reply src/base-converter.ts]: No reason that I can tell. I'll move this check a level up so that this method always performs the conversion its told to.

**Comment by FredKSchott** [comment]
@justinfagnani PTAL

**Pull request merged**